crates/oxc_semantic/src/builder.rs

```diff
@@ -100,6 +100,58 @@ pub struct SemanticBuilderReturn<'a> {
     pub errors: Vec<OxcDiagnostic>,
 }
 
+#[derive(Default, Debug)]
+pub struct Collector {
+    node: usize,
+    scope: usize,
+    symbol: usize,
+    reference: usize,
+}
+
+impl<'a> Visit<'a> for Collector {
+    #[inline]
+    fn enter_node(&mut self, _: AstKind<'a>) {
+        self.node += 1;
+    }
+    #[inline]
+    fn enter_scope(&mut self, _: ScopeFlags, _: &Cell<Option<ScopeId>>) {
+        self.scope += 1;
+    }
+    #[inline]
+    fn leave_node(&mut self, _: AstKind<'a>) {}
+    #[inline]
+    fn leave_scope(&mut self) {}
+
+    fn visit_binding_identifier(&mut self, _: &BindingIdentifier<'a>) {
+        self.symbol += 1;
+        self.node += 1;
+    }
+    fn visit_identifier_reference(&mut self, _: &IdentifierReference<'a>) {
+        self.reference += 1;
+        self.node += 1;
+    }
+    fn visit_jsx_identifier(&mut self, _: &JSXIdentifier<'a>) {
+        self.reference += 1;
+        self.node += 1;
+    }
+    fn visit_jsx_member_expression_object(&mut self, it: &JSXMemberExpressionObject<'a>) {
+        if let JSXMemberExpressionObject::MemberExpression(expr) = &it {
+            self.visit_jsx_member_expression(expr);
+        } else {
+            self.node += 1;
```

Comment:
Since you're counting `JSXIdentifier` as a reference, shouldn't this branch also have `self.reference += 1;`?

Reply:
https://github.com/oxc-project/oxc/blob/b4a8d675e0dd9999f2098000fecbb4330470d45e/crates/oxc_semantic/src/builder.rs#L1966-L1975

`JSXIdentifier` is not always a reference. you can see my latest [commit](https://github.com/oxc-project/oxc/pull/4328/commits/b4a8d675e0dd9999f2098000fecbb4330470d45e).

Reply:
Two things:

In that case shouldn't it call `self.visit_jsx_identifier`?

I don't think we necessarily need to make the counts 100% accurate. Over-estimating is fine. It may be that the cost of complicating `Collector` to make an accurate count (and so making it a bit slower) is worse than the cost of allocating too much space in the `Vec`s. It's *under*-estimating that we absolutely need to avoid.

Reply:
> Two things:
> 
> In that case shouldn't it call `self.visit_jsx_identifier`?

If I call the `visit_jsx_identifier`, the reference count will increase. But it's not a `reference`

Reply:
Unless I'm reading `reference_jsx_identifier` wrong, it *does* create a reference for a `JSXIdentifier` whose parent is a `JSXMemberExpressionObject`. `Some(AstKind::JSXMemberExpressionObject(_)) => {}` lets execution fall through to next block of code which creates the reference.

That would be correct behavior - in `<Foo.bar />`, `Foo` is a reference.

But this isn't easy to reason about. #3528 would make it much simpler.